src/Remapping/distributed_remapping.jl

```diff
@@ -144,65 +219,482 @@ function Remapper(target_hcoords, target_zcoords, space)
             global_elem_lidx[gidx] = lidx
         end
 
-        local_indices = map(local_target_hcoords) do hcoord
-            helem = Meshes.containing_element(horz_mesh, hcoord)
+        local_horiz_indices = map(local_target_hcoords) do hcoord
+            helem = Meshes.containing_element(horizontal_mesh, hcoord)
             return global_elem_lidx[horizontal_topology.orderindex[helem]]
         end
     end
 
+    local_horiz_indices = ArrayType(local_horiz_indices)
+
+    field_values_size = ntuple(_ -> Nq, num_hdims)
+    field_values = ArrayType(zeros(FT, field_values_size))
+
     # We represent interpolation onto an horizontal slab as an empty list of zcoords
-    isnothing(target_zcoords) && (target_zcoords = [])
+    if isnothing(target_zcoords) || target_zcoords == []
+        target_zcoords = nothing
+        vert_interpolation_weights = nothing
+        vert_bounding_indices = nothing
+        local_interpolated_values =
+            ArrayType(zeros(FT, size(local_horiz_indices)))
+        interpolated_values =
+            ArrayType(zeros(FT, size(local_target_hcoords_bitmask)))
+    else
+        vert_interpolation_weights =
+            ArrayType(vertical_interpolation_weights(space, target_zcoords))
+        vert_bounding_indices =
+            ArrayType(vertical_bounding_indices(space, target_zcoords))
+
+        # We have to add one extra dimension with respect to the bitmask/local_horiz_indices
+        # because we are going to store the values for the columns
+        local_interpolated_values = ArrayType(
+            zeros(FT, (size(local_horiz_indices)..., length(target_zcoords))),
+        )
+        interpolated_values = ArrayType(
+            zeros(
+                FT,
+                (size(local_target_hcoords_bitmask)..., length(target_zcoords)),
+            ),
+        )
+    end
 
     return Remapper(
         comms_ctx,
+        space,
         local_target_hcoords,
         target_zcoords,
         local_target_hcoords_bitmask,
-        interpolation_coeffs,
-        local_indices,
-        space,
+        local_horiz_interpolation_weights,
+        local_horiz_indices,
+        vert_interpolation_weights,
+        vert_bounding_indices,
+        local_interpolated_values,
+        field_values,
+        interpolated_values,
     )
 end
 
+function Remapper(
+    target_hcoords::AbstractArray,
+    target_zcoords::Union{AbstractArray, Nothing},
+    space::Spaces.AbstractSpace,
+)
+    Base.depwarn(
+        "The order of arguments in Remapper has changed. Please use Remapper(space, target_hcoords, target_zcoords) instead.",
+        :Remapper,
+    )
+    return Remapper(space, target_hcoords, target_zcoords)
+end
 
 """
-   interpolate(remapper, field; physical_z = false)
+    _set_interpolated_values!(remapper, field)
 
-Interpolate the given `field` as prescribed by `remapper`.
+Change the local state of `remapper` by performing interpolation of `Fields` on the vertical
+and horizontal points.
```

Comment:
Can we expand on what this is doing mathematically?

Reply:
Should we convert that comment to a doc string?

Reply:
I added references. The commend is too longwinded to be a useful docstring.